platform/kube/inject/inject.go

```diff
@@ -123,16 +170,62 @@ func GetMeshConfig(kube kubernetes.Interface, namespace, name string) (*proxycon
 	return &mesh, nil
 }
 
-func injectIntoPodTemplateSpec(p *Params, t *v1.PodTemplateSpec) error {
-	if t.Annotations == nil {
-		t.Annotations = make(map[string]string)
-	} else if _, ok := t.Annotations[istioSidecarAnnotationSidecarKey]; ok {
-		// Return unmodified resource if sidecar is already present or ignored.
-		return nil
+func injectRequired(namespacePolicy InjectionPolicy, objectMeta *metav1.ObjectMeta) bool {
+	var resourcePolicy string
+	if objectMeta.Annotations == nil {
+		resourcePolicy = istioSidecarAnnotationPolicyValueDefault
+	} else {
+		var ok bool
+		if resourcePolicy, ok = objectMeta.Annotations[istioSidecarAnnotationPolicyKey]; !ok {
+			resourcePolicy = istioSidecarAnnotationPolicyValueDefault
+		}
+	}
+
+	var required bool
+	switch namespacePolicy {
+	case InjectionPolicyOptIn:
+		if resourcePolicy == istioSidecarAnnotationPolicyValueForceOn {
+			required = true
+		}
+	case InjectionPolicyOptOut:
+		if resourcePolicy != istioSidecarAnnotationPolicyValueForceOff {
+			required = true
+		}
 	}
-	t.Annotations[istioSidecarAnnotationSidecarKey] = istioSidecarAnnotationSidecarValue
-	t.Annotations[istioSidecarAnnotationVersionKey] = p.Version
 
+	status, ok := objectMeta.Annotations[istioSidecarAnnotationStatusKey]
+
+	glog.V(2).Infof("Sidecar injection policy for %v/%v: namespace:%v resource:%v status:%q required:%v",
+		objectMeta.Namespace, objectMeta.Name, namespacePolicy, resourcePolicy, status, required)
+
+	if checkDeprecatedAlphaAnnotation {
+		if _, ok = objectMeta.Annotations[deprecatedIstioSidecarAnnotationSidecarKey]; ok {
+			return false
+		}
+	}
+
+	if !required {
+		return false
+	}
+
+	// TODO - add version check for sidecar upgrade
+
+	return !ok
```

Comment:
so the idea is that if the status annotation is present then there should be no injection?